### PR TITLE
Pmd7 upgrade 7.2.0 concurrent rules set 03

### PR DESCRIPTION
**Pull request by stokpop**
Notes:

* `NotProperlySynchronizingOnFieldWhileUsingGuardedBy.xml` has one failing unit test, has to do with `synchronized(this)` vs `GuardedBy("this")`, could not find easy fix yet
* found that there are (at least) 3 `GuardedBy` annotation implementation: should we match all three explicitly? Maybe match on name is easier in this case...
  * net.jcip.annotations.GuardedBy
  * javax.annotation.concurrent.GuardedBy
  * com.google.errorprone.annotations.concurrent.GuardedBy

### CODE REVIEW

File: src/main/resources/category/java/concurrent.xml [hunk 1]
```diff
@@ -822,44 +913,56 @@ public class Good {
...
+        or pmd-java:hasAnnotation('org.springframework.beans.factory.annotation.Value')
+        or (pmd-java:hasAnnotation('javax.inject.Inject') or pmd-java:hasAnnotation('jakarta.inject.Inject'))
+    )
+ ]
+(: a variable which should (potentially) be guarded, skip synchronized(this) vars :)
+//(AssignmentExpression/*[1]|VariableAccess[not(parent::SynchronizedStatement) and not(ancestor::SynchronizedStatement/ThisExpression)])[
+   (: used in a class with guarded-by literal *IS NOT* a field (should be private final) :)
+   @Name = ancestor::ClassDeclaration//FieldDeclaration[
+        ModifierList/Annotation[pmd-java:typeIs('net.jcip.annotations.GuardedBy') or pmd-java:typeIs('javax.annotation.concurrent.GuardedBy') or pmd-java:typeIs('com.google.errorprone.annotations.concurrent.GuardedBy')]//MemberValuePair/StringLiteral[
+             not(substring(@Image, 2, (string-length(@Image) - 2)) = ancestor::ClassDeclaration//FieldDeclaration/VariableDeclarator/VariableId/@Name)]
```
Comment (jborgers) [inline src/main/resources/category/java/concurrent.xml:957]: Kun je hier niet beter @LiteralText gebruiken? Gebruik van @Image wordt meen ik afgeraden. Dan heb je 1 character minder voor en achter. 

En misschien is gebruik van matches i.p.v. substring hier korter (maar wellicht trager)

File: src/main/resources/category/java/concurrent.xml [hunk 2]
```diff
@@ -822,44 +913,56 @@ public class Good {
...
+          ]/VariableDeclarator/VariableId/@Name
+    (: is defined as field :)
+    /ancestor::ClassDeclaration//FieldDeclaration/VariableDeclarator/VariableId/@Name
+    (: and *IS* in a synchronized block with the GuardedBy literal:)
+    and (concat('"', ancestor::SynchronizedStatement/VariableAccess/@Name, '"') =
+       ancestor::ClassDeclaration//FieldDeclaration[pmd-java:hasAnnotation('net.jcip.annotations.GuardedBy') or pmd-java:hasAnnotation('javax.annotation.concurrent.GuardedBy') or pmd-java:hasAnnotation('com.google.errorprone.annotations.concurrent.GuardedBy')]//MemberValuePair/StringLiteral/@Image)
+]
```
Comment (jborgers) [inline src/main/resources/category/java/concurrent.xml:964]: @LiteralText ?

**Review by jborgers — APPROVED** [review]
paar opmerkingen, wellicht nog wat aanpassen

**Comment by jborgers** [comment]
> Notes:
> 
> * `NotProperlySynchronizingOnFieldWhileUsingGuardedBy.xml` has one failing unit test, has to do with `synchronized(this)` vs `GuardedBy("this")`, could not find easy fix yet
> * found that there are (at least) 3 `GuardedBy` annotation implementation: should we match all three explicitly? Maybe match on name is easier in this case...
>   
>   * net.jcip.annotations.GuardedBy
>   * javax.annotation.concurrent.GuardedBy
>   * com.google.errorprone.annotations.concurrent.GuardedBy

I think if there is no reason to have confusion like with Lombok and Spring @Value, with different meanings, it is easier to just use the name and maybe add the fully qualifications in comments, in case confusion arises in the future.

File: src/main/resources/category/java/concurrent.xml (hunk 2 above)
Comment (jborgers) [reply src/main/resources/category/java/concurrent.xml:964]: @ConstValue

File: src/main/resources/category/java/concurrent.xml (hunk 1 above)
Comment (stokpop) [reply src/main/resources/category/java/concurrent.xml]: Replaced with @ConstValue, and Simplified

File: src/main/resources/category/java/concurrent.xml (hunk 2 above)
Comment (stokpop) [reply src/main/resources/category/java/concurrent.xml:965]: Replaced with @constvalue, and Simplified, also fixed all unit tests

**Review by jborgers — APPROVED** [review]

**Pull request merged**